hangman/lib/hangman/controller.ex

```diff
@@ -0,0 +1,77 @@
+defmodule Hangman.Controller do
```

Comment:
Wow. This is an amazing amount of work. I'm impressed.

However, I think there's a simpler way.

Hangman.new_game already returns the pid of the hangman game server. That pid is stored by human_player, and used to make moves. So there's no need for all the message passing: Erlang will do it for you, because when you call Hangman.make_move, it does a GenServer call to that pid, which gets sent to the remote process.

I'm not deducting anything—I just wanted to point it out.

In general, you probably don't want to be doing low level receive stuff like this.

Reply:
Yeah,I 100% understand what is wrong with my work and 100% agree with your comment, professor.

I just focused too much on my initial ideas about the overall high-level structures of the application. And I forgot that in the Erlang\Elixir world, remote communication between processes is such a easy and convenient thing. I just stayed quite a long time in the Java world with the old-fashion C/S communication. All the thing I know about threads in Java(I guess they act like processes in Erlang...) is how to implement some concurrent thread tasks and never imagine their communication XD.

I were at first thinking that the highest supervisor should supervise a worker(the controller in my case) and a secondary supervisor. And the worker should listen to business requests from clients and the secondary supervisor supervises genservers as game servers. And I were thinking that the server worker should be the only interface visible to the clients, and during communication a client shouldn't know any details about the server,e.g. the genserver pid. A client just tells the server what he wants and then waiting for the server waitress to serve him the food.

I agree that this design is rather unnecessary and may be even wrong in the Erlang world. Some low-level direct communication as you mentioned between processes will be a much simpler way.  Elixir has been changing many old concepts and patterns in my mind and I really enjoy it. I really appreciate your suggestions and thank you professor!